src/burnchains/bitcoin/blocks.rs

```diff
@@ -17,119 +17,130 @@
 
-// IPC messages between threads
-pub struct IPCHeader {
-    pub height: u64,
-    pub header: LoneBlockHeader
+use util::hash::to_hex;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct BitcoinHeaderIPC {
+    pub block_header: LoneBlockHeader,
+    pub block_height: u64
 }
 
-pub struct IPCBlock {
-    pub height: u64,
-    pub header: LoneBlockHeader,
-    pub block: PeerMessage
+impl BurnHeaderIPC for BitcoinHeaderIPC {
+    type H = LoneBlockHeader;
+
+    fn header(&self) -> LoneBlockHeader {
+        self.block_header.clone()
+    }
+
+    fn height(&self) -> u64 {
+        self.block_height
+    }
 }
 
-pub struct BitcoinBlockDownloader {
-    headers_path: String,
-    cur_request: Option<Arc<IPCHeader>>,
-    network_id: BitcoinNetworkType,
+#[derive(Debug, Clone, PartialEq)]
+pub struct BitcoinBlockIPC {
+    pub header_data: BitcoinHeaderIPC,
+    pub block_message: PeerMessage
+}
+
+impl BurnBlockIPC for BitcoinBlockIPC {
+    type H = BitcoinHeaderIPC;
+    type B = PeerMessage;
+
+    fn header(&self) -> BitcoinHeaderIPC {
+        self.header_data.clone()
+    }
+
+    fn height(&self) -> u64 {
+        self.header_data.height()
+    }
+
+    fn block(&self) -> PeerMessage {
+        self.block_message.clone()
+    }
+}
 
-    pub chan_in: Option<Receiver<Arc<IPCHeader>>>,
-    pub chan_out: Option<SyncSender<Arc<IPCBlock>>>,
-    pub thread: Option<JoinHandle<()>>
+pub struct BitcoinBlockDownloader {
+    cur_request: Option<BitcoinHeaderIPC>,
+    cur_block: Option<BitcoinBlockIPC>,
+    indexer: Option<BitcoinIndexer>
 }
 
 pub struct BitcoinBlockParser {
     network_id: BitcoinNetworkType,
-    magic_bytes: MagicBytes,
-
-    pub chan_in: Option<Receiver<Arc<IPCBlock>>>,
-    pub chan_out: Option<SyncSender<Arc<BurnchainBlock<BitcoinAddress, BitcoinPublicKey>>>>,
-    pub thread: Option<JoinHandle<()>>
+    magic_bytes: MagicBytes
 }
 
 impl BitcoinBlockDownloader {
-    pub fn new(headers_path: &str, start_block: u64, end_block: u64, network_id: BitcoinNetworkType) -> BitcoinBlockDownloader {
+    pub fn new(indexer: BitcoinIndexer) -> BitcoinBlockDownloader {
         BitcoinBlockDownloader {
-            headers_path: headers_path.to_owned(),
             cur_request: None,
-            network_id: network_id,
-            chan_in: None,
-            chan_out: None,
-            thread: None
+            cur_block: None,
+            indexer: Some(indexer)
         }
     }
 
-    // TODO: connection methods and thread start
-    // TODO: receive a block and send it off
-
-    /// Go get all the blocks.
-    /// keep trying forever.
-    pub fn run(&mut self, indexer: &mut BitcoinIndexer) -> Result<(), btc_error> {
-        return indexer.peer_communicate(self);
+    pub fn run(&mut self, header: &BitcoinHeaderIPC) -> Result<BitcoinBlockIPC, btc_error> {
+        self.cur_request = Some((*header).clone());
+        
+        // let mut indexer = self.indexer.dup();
+        let mut indexer = self.indexer.take().unwrap();
```

Comment:
Do you do this so that only thread will have access to the indexer at a time?

Reply:
It's a hack around the borrow checker -- this prevents a mutable reference cycle.  Without this, you'd have to call something like `self.indexer.unwrap().peer_communicate(self, false)`, which borrows `self` mutably twice.  By calling `self.indexer.take()`, we're removing the reference to the indexer from `self` and putting it into a temporary variable.  Then, the borrow checker has no problem with `indexer.peer_communicate(self, false)`.

Reply:
Gotcha